.github/workflows/stack.yml

```
@@ -14,27 +14,37 @@ jobs:
           - "8.10.7"
 
     steps:
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v3
         with:
           submodules: true
 
-      - uses: haskell/actions/setup@v1
-        name: Setup Stack
+      - name: Setup Stack
+        uses: haskell/actions/setup@v2
         with:
           enable-stack: true
           stack-version: "latest"
           cabal-version: ${{ matrix.cabal }}
 
-      - uses: pavpanchekha/setup-z3@1.2.2
-        name: setup z3
+      - name: setup z3
+        uses: pavpanchekha/setup-z3@1.2.2
         with:
           version: "4.8.7"
 
-      - uses: actions/cache@v2
-        name: Cache ~/.stack
+      - name: Cache ~/.stack
+        uses: actions/cache@v3
         with:
           path: ~/.stack
-          key: ${{ runner.os }}-${{ matrix.ghc }}-stack
+          key: ${{ runner.os }}-stack-global-${{ hashFiles('stack.yaml') }}-${{ hashFiles('package.yaml') }}
+          restore-keys: |
+            ${{ runner.os }}-stack-global-
+
+      - name: Cache .stack-work
+        uses: actions/cache@v3
+        with:
+          path: .stack-work
+          key: ${{ runner.os }}-stack-work-${{ hashFiles('stack.yaml') }}-${{ hashFiles('package.yaml') }}-${{ hashFiles('**/*.hs') }}
+          restore-keys: |
+            ${{ runner.os }}-stack-work-
```

Maybe
```suggestion
            ${{ runner.os }}-stack-work-${{ hashFiles('stack.yaml') }}
            ${{ runner.os }}-stack-work-
```
Otherwise, the retrieved cache may not match the needed dependencies.

Also, isn't adding `-${{ hashFiles('**/*.hs') }}` too specific? I doubt that we will get exact key matches if we keep that.

---

This I took as-is from https://github.com/actions/cache/blob/main/examples.md#haskell---stack. Thanks for checking it (I didn't :-(). I'll tweak it for our use and to be more like the cabal caching.